util: Added cli tool for gem5 resources

Harshil2107 opened the pull request:
- create_resource is functional but need polishing

- Validate and Update resource are work in progress.

Change-Id: I6ec65ccef72257dd9ea6e5e83019306834be7367

Harshil2107 committed:
util: Added cli tool for gem5 resources

- create_resource is functional but need polishing

- Validate and Update resource are work in progress.

Change-Id: I6ec65ccef72257dd9ea6e5e83019306834be7367

Harshil2107 commented:
Note: Cli will not work without the URI as an env variable

Harshil2107 committed:
stdlib: Added skeleton of Updater

Change-Id: Id77ae0654347e501451fbd5fd607ad6a0b668504

BobbyRBruce committed:
util: Impr. dir structure and build sys for gem5-res-man

1. Sources moved to "src" directory.
2. `gem5_resources_manager` Python module created in "src" to contain
   sources as a module (note includsion of "__init__.py" so code may be
   imported with `from gem5_resources_manager import x`.
3. `__main__.py` included to give logical place for `main` entry point.
4. "pyproject.toml" added. This allows for automated dependency getting
   (just Pymongo for now), building, and installation.

Run "pip install ." to install the tool as `gem5_resources_manager`.
This can then be executed, via the command line, with
`gem5-resources-manger`.

(If you use `pip install -e .` then you can work on the python files and
they should update `gem5-resources-manager` without re-building).

Change-Id: I0fbe7ee27847a476db2c4d28be607bd68ba3d13a

BobbyRBruce committed:
util: Add .gitignore

Change-Id: Ie2754345e9c7914cd876027a37ca63918e358577

BobbyRBruce committed:
util: Mv 'Updated-gem5-cli' to 'gem5-resources-manager-v2'

This is less confusing

Change-Id: I4b743a8f44ba5fd5095a788881e315ba59bd27a1

BobbyRBruce committed:
misc: Merge branch 'develop' into resources-tooling

Change-Id: Ib673db6f251884a8b1e38d7c03ba40efa1953e73

BobbyRBruce committed:
util: Move subtool "Creator" into it's own submodule

This uses OO design to create an interface (`AbstractSubtool`) for which
this tool may be extended, modified, and kept organized.

As an example i have half-implenented the "Creator" subtool.

**Important notes:**

1. This has not been tested, it's only an example.
2. I have removed the other subtools just to clear up the space to show
   this example.

Change-Id: Ia0ff3f31159c971fade7594708be3b7db4707081

BobbyRBruce committed:
util: Create resource api and DummyResource stub

Here I've created a `AbstractDataSource` class. I use this to define the
API i need an expect for _any_ gem5 resources datasource I'm going to
connect to. With that defined I create a stub `DummyDatabase` to
imlement this and check it's functionality. There's no need at this
stage to worry about database interaction, or even writing out to file.

The layout, structure, and design methodology is what i'm trying to get
across. The API is likely too basic for our needs and nothing is
complete (I didn't even test it), but it should give an idea of how
things can be laid out in a cleaner, more managable way.

Change-Id: I1b7511c3c55946c87cf381c21b21fe28dd487c06

BobbyRBruce commented:
I think the direction you were heading in your old design was too haphazrd and was going to make it hard for you to maintain, expand, test, and improve going forward. 

I pushed some patches which I hope will  show a slightly better approach. Nothing I've submitted is complete, or even what you may need to fully complete the task, but it just serves as an example/tutorial of a better way to do things.

I'd advise going through each commit one-by-one to see the development.

As a very high-level note I'll say that you should think about creating a design with is easy to extend without modification. If i want to create a new datasource I should be able to create a new class which implements an API, then just plug it in. If modification is necessary then the modification should only affect that one thing, in one place. If the schema changes I want to have just one place I change it in this tool, nothing more.

Harshil2107 commented:
Thank you for the feedback. I will look into the suggested changes.

Harshil2107 committed:
util: created validators and json data source

Change-Id: If5c592d67a5626943de3fbd1e94e2cb26215c504

BobbyRBruce committed:
util: Cleanup entry points

* Get rid of "__main__.py" for gem5_resource_manager.

We now declare `main` functions in "__init__.py" files.

Change-Id: Idde0301b60125a45bc4b7a5096fdd052bbe4cefc

BobbyRBruce committed:
util: Comment out Loader import

This doesn't exist.

Change-Id: Ie49701f2242c476ffaaadc9d990c1ae859fbe145

BobbyRBruce committed:
util: Add "DummyDatabase" to __init__.py

This means imporst such as `from data_source import DummyDatabase`
function as intended.

Change-Id: I0b78f7fedcf2a338e32f673f53fe4565d5726cbf

BobbyRBruce committed:
util: Fix argparsing for subtool

Change-Id: I16d018481691b6de0975f605e32b4ee9621bf339

BobbyRBruce commented:
I'm not sure exactly what you were doing but I fixed the bugs Iw as running into with this workflow:

```sh
pip install -e . # Installs the tool.
gem5-resources-manager # Run the tool.
```

Harshil2107 committed:
util: added resource creator skeleton

Change-Id: I641508579fa1fbd9103f579c9273dc2066f33e1c

Harshil2107 committed:
util: Added updater as a subtool

Change-Id: Ic6998c91f436d2d8078fd91929cae07761e336eb

Harshil2107 committed:
util: added skeleton for recursive updates

Change-Id: I7886fa96e6c9e2ffb7bf2c60fe20c9907e636866

Harshil2107 committed:
util: reformated updater

- auto updating dependent resources is still wip

Change-Id: I396a7028e307e1be630777b610e3f74a6d22c811

Harshil2107 committed:
util: Added support for Atlas database

- you currently can not test it as the read only uri is not generated

Change-Id: Ic2755b44f39df49530aca590722951fc0097e947

BobbyRBruce commented on util/gem5-resources-manager-v2/resource_demo.json:
Format this file with new-lines and intends.

BobbyRBruce commented on util/gem5-resources-manager-v2/src/gem5_resources_manager/creator/creator.py:
```diff
@@ -0,0 +1,154 @@
+            "--handle-url",
+            "-u",
+            action="store_true",
+            help="automatically fill in the is_zipped, is_tarred, and md5sum fields if the url field has a link to the source repository.",
```
Remember to break up your strings if they get too long.

BobbyRBruce commented on util/gem5-resources-manager-v2/test-scritp.py:
I really really really REAALY want tests for this, but can you write some Pyunit tests instead?

This is not fun to read and non-standard.

BobbyRBruce commented on util/gem5-resources-manager-v2/src/gem5_resources_manager/updater/updater.py:
```diff
@@ -0,0 +1,126 @@
+
+        print("Updated resource:", json.dumps(updated_resource, indent=4))
+
+    def get_dependent_resources(self, resource, dependent_resources):
```
There's a loy of functions in this PR that don't have typing information. This include return type info.

Pleaes try to get into the habit of adding these. In this case it's hard to know what `dependent_resources` is supposed to be. A set of resources, i'd guess?

BobbyRBruce commented on util/gem5-resources-manager-v2/src/gem5_resources_manager/resource_cli.py:
```diff
@@ -0,0 +1,30 @@
+    )
+
+    datasource.open()
+    subtool.execute(args=args, data_source=datasource)
```
This is cause of your bug. At least in part. What were you trying to do here? `subtool` is just the variable that holds the value in the for iteration. Once that's over it's just going to be the last thing set.

You've changed this code a lot, when i handed it to you it looked like this:

```py
subparserer = parser.add_subparsers()

     parser.add_argument("obj")
     for subtool in subtools:
         subtool_parser = subtool.__class__.get_arg_parser()
         subtool_parser.set_defaults(obj=subtool)
         subparserer.add_parser(subtool.__class__.get_arg_parser())

     args = parser.parse_args()

     subtool = args.obj
```

The key here is we add the argument `obj` to the parser, then for each sub parser we override `obj`  with the `AbstractSubtool` that `subtool_parser` is associated with. Given we only ever use one subtool, which select at the start, this `obj` value will only be overriden by the subtool you choose. (Note `obj` is a bad name. Here is is clearly a class).

After this we can do something like `subtool().execute()` or whatever it can do. You should probably try changing this back.

BobbyRBruce commented on util/gem5-resources-manager-v2/src/gem5_resources_manager/abstract_subtool.py:
```diff
@@ -0,0 +1,11 @@
+
+
+class AbstractSubtool(ABC):
+    def get_arg_parser() -> ArgumentParser:
```
Due to your redesign, this is no longer the signature of this function.

BobbyRBruce commented on util/gem5-resources-manager-v2/src/gem5_resources_manager/abstract_subtool.py:
```diff
@@ -0,0 +1,11 @@
+    def get_arg_parser() -> ArgumentParser:
+        raise NotImplementedError
+
+    def execute(args: Namespace, data_source: AbstractDataSource):
```
Same story here: this isn't a member function, it's a class function.

BobbyRBruce commented on util/gem5-resources-manager-v2/src/gem5_resources_manager/creator/creator.py:
```diff
@@ -0,0 +1,154 @@
+        )
+
+    def execute(self, args: Namespace, data_source: AbstractDataSource):
+        self.data_source = data_source
+        print("Creator.execute()", args.category)
+        self.create_resource_json(args)
```
This is messy. Why are we calling execute, setting the member variable `data_source` then calling the function `create_resource_json` with `args`. Then in `create_resource_json` using both `args` and `self.data_source`. If you think it through, why not just move `create_resource_json` into `execute`? And why have member variables, or even have these as member functions?

Do we need a `Creator` Object to hold ata and state information? Isn't this a better case for functional programming?

BobbyRBruce commented on util/gem5-resources-manager-v2/src/gem5_resources_manager/creator/creator.py:
```diff
@@ -0,0 +1,154 @@
+        resource = {}
+        required, optional = self.data_source.get_fields(args.category)
+        populated_fields = ast.literal_eval(args.field_entries)
+        resource["category"] = args.category
+        del required["category"]
```
Here you are just setting then deleting a variable

BobbyRBruce commented on util/gem5-resources-manager-v2/src/gem5_resources_manager/creator/creator.py:
```diff
@@ -0,0 +1,154 @@
+            self.data_source.validate_resource(resource)
+        self.data_source.create_new_entry(resource)
+
+    def enter_fields(self, fields, resource, populated_fields, args, is_optional=False):
```
There's so much confusion here. What information do we want? I think

* The list of fiends for each category.
* For each field a to know if ti's optonal.
* The "type" fo the field (its "schema").
* If we're running in `verbose` mode.
* If `is_optional`.

So, how about you create a new class called `FieldType` which encapulates all the schema information about a field. This woulde include info on whether its optinal, andprovide the field's accepted value types as well as a validator. We can then have a function which returns a List of `Field` objects for that category

Also, let's rename `is_optional` to `ask_on_optionmal_param`. `is_optional` doesn't have any meaning to me.

 Here's the basic pseudo-coee:



```py
field_types = get_field_types(category="binary")

for field_type in field_types:
    if field_type.is_optional():
        if not ask_on_optional_param:
            continue # No inclusion of optional parameters.
        # Here you do "choice logic" asking if they want to 
        # set the parameter. If "no" `continue` called.
    
    val = None
    while not val:
        print(f"The field {field_type.get_name()} expects a value of {field_type.get_value_type_as_str()}")
        val = input('bla bla')
        if not field_type.is_valid_value(val):
            val = None
            print("Incorrect value type!")
 
    # Set the value to the resource here
```

And we've discussed thia many, many times now, and I've still got no movement on it: Can we please cleanup this default value/None value/field missing mess. I really think it'd help a lot if we implemented the following rule:

1. A missing field is consider to have a value of `None`.
2. In line with 1. an "Optional" field is a field that can be `None` or a value.
3. We drop this default value thing. If a field is optional and not set by the user, it's stored as None. If it's not optional then the user is promted for a value which must be given, with this value being valid in accordance to the schema.

Harshil2107 replied:
Here is my interpretation of this:
- If a field is missing that is the key does not exist then it means that the resource doesnot have the field
- Currently the fields cannot be None or Null but if they are lets say "" or [] it means that the resource has the field but the value is an empty string or empty array.

BobbyRBruce commented on util/gem5-resources-manager-v2/src/gem5_resources_manager/creator/creator.py:
```diff
@@ -0,0 +1,154 @@
+        is_zipped = zipfile.is_zipfile("temp")
+        is_tarred = tarfile.is_tarfile("temp")
+        # if url ends with .tar or .tar.gz, then it is tarred
+        if url.endswith(".tar"):
+            is_tarred = True
+        if url.endswith(".tar.gz"):
+            is_tarred = True
+            is_zipped = True
```
Don't do it this way.  You can't infer file type by extension. Use Python's tar and gzip library

BobbyRBruce commented on util/gem5-resources-manager-v2/src/gem5_resources_manager/creator/creator.py:
```diff
@@ -0,0 +1,154 @@
+        populated_fields["is_zipped"] = is_zipped
+        populated_fields["is_tar_archive"] = is_tarred
+        populated_fields["md5sum"] = hashlib.md5(
+            file_as_bytes(open("temp", "rb"))
```
This won't work for directories. Import the code from md5.py

BobbyRBruce commented on util/gem5-resources-manager-v2/src/gem5_resources_manager/creator/creator.py:
```diff
@@ -0,0 +1,154 @@
+                self.handle_url(resource[field], populated_fields)
+
+    def handle_url(self, url, populated_fields):
+        urllib.request.urlretrieve(url, "temp")
```
use "pathutil" to automatically create a tempfile

BobbyRBruce commented on util/gem5-resources-manager-v2/src/gem5_resources_manager/data_source/atlas_data_source.py:
```diff
@@ -0,0 +1,170 @@
+        pass
+
+    def resource_exists(
+        self, resource_id: str, resource_version: Optional[str] = "1.0.0"
```
Why does the `resource_version` have a default?

BobbyRBruce commented on util/gem5-resources-manager-v2/src/gem5_resources_manager/data_source/atlas_data_source.py:
```diff
@@ -0,0 +1,170 @@
+        )
+        json_resource = json_util.dumps(resource)
+        res = json.loads(json_resource)
+        if res == []:
+            return False
+        return True
+
```
```suggestion
        return bool(res)

```

BobbyRBruce commented on util/gem5-resources-manager-v2/src/gem5_resources_manager/data_source/atlas_data_source.py:
```diff
@@ -0,0 +1,170 @@
+        self, resource_id: str, resource_version: str
+    ) -> ResourceCategory:
+        resource = (
+            self.collection.find(
+                {
+                    "id": resource_id,
+                    "resource_version": resource_version,
+                },
+                {"_id": 0},
+            )
```
You're writing this over and over. Make a private function calkled "get_resource" so you just write it the once.

BobbyRBruce commented on util/gem5-resources-manager-v2/src/gem5_resources_manager/data_source/atlas_data_source.py:
```diff
@@ -0,0 +1,170 @@
+            raise Gem5DataSourceEntryNotFound(resource_id)
+        return res[0]
+
+    def get_resource_from_file(self, file):
+        # append resource to resources.json
+        resources = None
+        if os.path.exists(file):
+            with open(file) as infile:
+                resources = json.load(infile)
+        return resources
```
WHy is this here? This has nothing to do with the Atlas DB.

BobbyRBruce commented on util/gem5-resources-manager-v2/src/gem5_resources_manager/data_source/dummy_database.py:
I didn't bother to review this file RN.

BobbyRBruce commented on util/gem5-resources-manager-v2/src/gem5_resources_manager/data_source/json_data_source.py:
I didn't bother to review this file RN.

BobbyRBruce commented on util/gem5-resources-manager-v2/src/gem5_resources_manager/schemas/json_validator.py:
I won't review this on this opas but i'll say you **need** to add comments here to explain your logic. This code smells. Too much hard coding, if statements, random deletion of fields. You need to comminicate the flow of logic here.

Harshil2107 committed:
util: Added auto update for related resources

Change-Id: I43e4d28f8626e4e25af987c05a76cf6833951af7

BobbyRBruce commented:
@Harshil2107 : I'll let you reopen this if you feel differently but I think we'll never get this in and should just close this and call it a dead end project (we can always change our mind later). If we didn't need it in the last 6 months, I doubt we'll ever need it. 

You may want to see if there's anything here worth salvaging given your work on gem5-resources in the past few months. Probably not of all of it, but maybe some? It's up to you. If you still feel like this may be of use at some point but you won't work on it anytime soon, create an issue explaining the idea a bit and link to this so someone else could theoretically have a go

Harshil2107 commented:
I moved away from needing a cli, as I feel that I only have 2 use cases that are prettry far away.

1: Just add of update one (or a few very different ones) resource for that the current resource manager works

2: To add or update a bunch of resources and in most cases these are unique among each change so I think just writing a script to create the JSON is faster.

I did push some helpful scripts to gem5/util if I was using the same ones a decent amount.
